<a href="https://colab.research.google.com/github/thaisNY/Soul-Code/blob/main/thais_soulcode_proximos_passos_em_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Próximos Passos com Python**

Este notebook é uma continuação dos primeiros passos em Python. Aqui você vai aprofundar seus conhecimentos com conceitos que programadores profissionais usam no dia a dia do trabalho.

## **O que é uma coleção**

Em Python, uma **coleção** é uma estrutura usada para **guardar vários valores dentro de uma única variável**, em vez de criar uma variável para cada valor.

Ela serve para **organizar, agrupar e trabalhar com conjuntos de dados relacionados**.

### Pensando de forma simples

Ao invés de fazer isso:

In [ ]:
a = 1
b = 2
c = 3

Você usa uma **coleção**:

In [ ]:
numeros = [1, 2, 3]

Uma única variável passa a representar **vários valores**.

### O que uma coleção permite fazer

Coleções permitem:

* armazenar **vários elementos**
* percorrer os dados com `for`
* verificar se algo existe (`in`)
* aplicar funções como `len()`, `max()`, `min()`
* organizar dados de formas diferentes, dependendo do tipo

### Tipos comuns de coleções em Python

Cada coleção tem um comportamento específico:

* **list** → ordenada, mutável, aceita repetição
* **tuple** → ordenada, não pode ser alterada
* **set** → não ordenada, não aceita repetição
* **dict** → pares de chave e valor

Todas são coleções porque **guardam vários dados juntos**, mas cada uma resolve um tipo de problema diferente.

### **Tuplas - A lista que não muda**

Uma **tupla** é uma estrutura de dados:

* Ordenada
* Imutável
* Pode armazenar múltiplos valores

Pense numa tupla como uma caixa lacrada: você pode ver o que tem dentro, mas não pode mudar o seu conteúdo.

In [ ]:
tupla = (1, 2, 3)
print(tupla)

### **Prática**

Crie uma tupla com três números e imprima

**Dica:** use parênteses `( )`

In [ ]:
# Escreva seu código aqui
tupla_frutas = ("uva", "limao", "pera")
print(tupla_frutas)

('uva', 'limao', 'pera')


### **Diferença entre Tupla e Lista**

* Listas são mutáveis
* Tuplas são imutáveis

In [ ]:
lista = [1, 2, 3]
tupla = (1, 2, 3)

# Lista é mutável
lista[0] = 100
print("Lista após modificação:", lista)

# Tupla é imutável
tupla[0] = 100  # Isso gera TypeError
print("Tupla após modificação:", tupla)

Lista após modificação: [100, 2, 3]


TypeError: 'tuple' object does not support item assignment

### **Prática**

Crie uma lista e uma tupla com os mesmos valores

Tente modificar o primeiro elemento de cada uma

**Dica:** observe o erro gerado na tupla

In [ ]:
# Escreva seu código aqui

pratos = ("sushi", "hamburguer", "carne")
print(pratos[0])
print(pratos[1])
print(pratos[2])

pratos[0] = "pasta"
print(pratos[0])

sushi
hamburguer
carne


TypeError: 'tuple' object does not support item assignment

### **Criando Tuplas**

#### Tupla comum

In [ ]:
t1 = (10, 20, 30)

#### Tupla com um elemento

In [ ]:
t2 = (5,)

#### Tupla sem parênteses

In [ ]:
t3 = 1, 2, 3

### **Prática**

Crie:
1. Uma tupla com um único valor
2. Uma tupla sem usar parênteses

**Dica:** a vírgula é obrigatória para tuplas de um elemento

In [ ]:
# Escreva seu código aqui
t1 = 3
t2 = 1, 2, 3
print(t1)
print(t2)

3
(1, 2, 3)


### **Tipos de Dados em Tuplas**

Tuplas podem conter vários tipos de dados.

In [ ]:
t = (1, "Eu sou incrível", 3.5, True)
print(t)

(1, 'Eu sou incrível', 3.5, True)


### **Prática**

Crie uma tupla com pelo menos 4 tipos de dados diferentes

Use um loop para imprimir o tipo de cada elemento

**Dica:** use `type()`

In [ ]:
# Escreva seu código aqui
tupla_exibida = "teste", False , 21.21 , 12

for i in range(len(tupla_exibida)):
  print(tupla_exibida[i])

teste
False
21.21
12


### **Acessando Elementos (Indexação)**

In [ ]:
t = (10, "Luffy", 5.4)

# Acesso por índice positivo
print(f"Primeiro elemento: {t[0]}")
print(f"Segundo elemento: {t[1]}\n")

# Acesso por índice negativo
print(f"Último elemento: {t[-1]}")
print(f"Penúltimo elemento: {t[-2]}")

Primeiro elemento: 10
Segundo elemento: Luffy

Último elemento: 5.4
Penúltimo elemento: Luffy


### **Prática**

Crie uma tupla com 5 elementos

Imprima o primeiro, o terceiro e o último

**Dica:** índices começam em 0

### **Fatiamento (Slicing)**

In [ ]:
t = (1, 2, 3, 4, 5)
print(f"{t[1:4]}\n")  # Retorna os elementos do índice 1 até o 3 (o 4 não entra)

computacao = (
    "Ada Lovelace",
    "Alan Turing",
    "Grace Hopper",
    "Donald Knuth",
    "Linus Torvalds"
)

print(computacao[1:4])  # Retorna do segundo ao quarto nome (índice 1 ao 3, fim exclusivo)

(2, 3, 4)

('Alan Turing', 'Grace Hopper', 'Donald Knuth')


### **Prática**

Use slicing para:
1. Pegar os dois primeiros elementos
2. Pegar os dois últimos

**Dica:** use índices negativos

### **Operações com Tuplas**

#### Concatenação

In [ ]:
(1, 2) + (3, 4)

(1, 2, 3, 4)

#### Repetição

In [ ]:
(1, 2) * 3

(1, 2, 1, 2, 1, 2)

### **Pense**

Como você explicaria o que aconteceu em "Concatenação" e "Repetição"?

### **Prática**

Crie duas tuplas

Concatene elas

Repita uma delas 2 vezes

**Dica:** use os operadores `+` e `*`

In [ ]:
(1,2,3) + (4,5,6)

(1, 2, 3, 4, 5, 6)

In [ ]:
(1,2,3) * 2


(1, 2, 3, 1, 2, 3)

### **Funções Úteis com Tuplas**

#### **Tupla com booleanos**

In [ ]:
bools = (True, False, True)

print(len(bools))
# Retorna 3 porque a tupla possui três valores

print(max(bools))
# Retorna True porque True é tratado como 1 e False como 0

print(min(bools))
# Retorna False porque False equivale a 0, que é menor que 1

print(sum(bools))
# Retorna 2 pois True vale 1 e False vale 0

3
True
False
2


#### **Tupla com números e booleanos**

In [ ]:
mistos = (10, True, 5, False)

print(len(mistos))
# Retorna 4 pois existem quatro elementos na tupla

print(sum(mistos))
# Retorna 16 pois True soma 1 e False soma 0 aos números

print(max(mistos))
print(min(mistos))
# IMPORTANTE: Não é recomendado comparar números e booleanos juntos, pois apesar de funcionar, pode gerar confusão na leitura do código


4
16
10
False


#### **Tupla com strings e booleanos**



In [ ]:
dados = ("Python", True, "Java")

print(len(dados))
# Retorna 3 porque a tupla contém três elementos

 #print(max(dados))
 #print(min(dados))
# IMPORTANTE: Isso gera erro, pois Python não consegue comparar strings com booleanos

3


#### **Tupla com apenas um elemento**

In [ ]:
unico = (True,)

print(len(unico))
# Retorna 1 porque a tupla possui apenas um elemento

print(sum(unico))
# Retorna 1 pois True equivale a 1

1
1


### **Prática**

* Crie uma tupla apenas com valores booleanos
  * Use len, max, min e sum
  * Observe os resultados e tente explicá-los

In [ ]:
# Escreva seu código aqui
tupla_boolean = (True, True, False, False, True)
print(len(tupla_boolean)) #5
print(max(tupla_boolean)) #True
print(min(tupla_boolean)) #False
print(sum(tupla_boolean)) #3

5
True
False
3


### **Prática**

* Crie uma tupla com números e booleanos
  * Teste sum e depois tente max ou min
  * Observe o comportamento

In [ ]:
# Escreva seu código aqui

numeros = (10, 2 , 100, False, 1 , True)
print(sum(numeros)) #114
print(max(numeros)) #100

114
100


### **Prática**

* Crie uma tupla com strings
  * Use len, max e min
  * Observe a ordem usada na comparação

In [ ]:
# Escreva seu código aqui

tupla_strings =  ("eu", "tu", "ele")
print(len(tupla_strings))
print(max(tupla_strings))
print(min(tupla_strings))  #python segue a ordem do alfabeto para determinar max e min

3
tu
ele


### **Métodos de Tuplas**

* **`count(valor)`**
  Retorna **quantas vezes** um determinado valor aparece dentro da tupla.
  É útil para verificar repetições, já que tuplas podem conter valores duplicados.

* **`index(valor)`**
  Retorna o **índice da primeira ocorrência** do valor informado na tupla.
  Se o valor não existir, ocorre um erro (`ValueError`).



In [ ]:
t = (1, 2, 2, 3)

print(t.count(2))
# Saída: 2
# O método count conta quantas vezes o valor 2 aparece na tupla. Como o número 2 aparece duas vezes, o resultado é 2.

print(t.index(3))
# Saída: 3
# O método index retorna a posição (índice) da PRIMEIRA ocorrência do valor informado. O número 3 está na posição 3 da tupla.

### **Prática**

Crie uma tupla com números repetidos

Conte quantas vezes um número aparece

Descubra o índice de um valor

**Dica:** Use `count()` e `index()`

In [ ]:
# Escreva seu código aqui
t = (10,10,5,1,10,2,1,2)
print(t.count(10)) #3
print(t.index(2)) #5 mostra o indice do primeiro

3
5


### **Desempacotamento de Tuplas**

In [ ]:
t = (10, 20)

a, b = t

print(a)
# Saída: 10
# O primeiro valor da tupla é atribuído à variável a.

print(b)
# Saída: 20
# O segundo valor da tupla é atribuído à variável b.

### **Prática**

Crie uma tupla com 3 valores

Desempacote em três variáveis

**Dica:** O número de variáveis deve bater com a tupla

In [ ]:
# Escreva seu código aqui

### **Desempacotamento com `*`**

No desempacotamento com `*`, uma variável pode receber vários valores de uma tupla ao mesmo tempo, armazenando-os em uma lista.

In [ ]:
t = (1, 2, 3, 4)

a, *b = t

print(a)
# Saída: 1
# O primeiro valor da tupla é atribuído à variável a.

print(b)
# Saída: [2, 3, 4]
# O * faz com que todos os valores restantes sejam agrupados em uma lista.

### **Prática**

Separe o primeiro elemento e o restante usando *

**Dica:** Apenas uma variável pode usar `*`

In [ ]:
# Escreva seu código aqui

### **Tuplas em Loops**

Observe o exemplo abaixo:
* A tupla t contém três valores: 1, 2 e 3.
* O for percorre a tupla elemento por elemento, na ordem em que eles estão.
* A variável item recebe um valor da tupla por vez a cada repetição do loop.
* Em cada iteração, print(item) exibe o valor atual.


In [ ]:
t = (1, 2, 3)
for item in t:
    print(item)

### **Prática**

Percorra uma tupla e imprima cada valor

**Dica:** Funciona igual a listas

In [ ]:
# Escreva seu código aqui

### **Tuplas Aninhadas**

* t é uma tupla de tuplas (tupla aninhada).
* t[0] acessa a primeira tupla interna, que é (1, 2).
* t[0][1] acessa o segundo elemento dessa tupla interna, que é 2.

*Primeiro índice escolhe qual tupla, o segundo índice escolhe qual valor dentro dela.*

In [ ]:
t = ((1, 2), (3, 4))
print(t[0][1])

### **Prática**

Crie uma tupla de tuplas

Acesse um valor interno

**Dica:** Use dois índices

In [ ]:
# Escreva seu código aqui

### **Tuplas como Chaves de Dicionário**

Um **dicionário** funciona como um sistema de **localização rápida**:
dada uma chave, o Python precisa encontrar o valor **sempre no mesmo lugar**.

Para isso, o Python cria uma espécie de **assinatura interna** da chave no momento em que ela é usada.

---

### O que o Python espera de uma chave?

* Que ela **não mude de forma**
* Que ela **represente sempre a mesma coisa**
* Que o Python consiga **reconhecê-la exatamente igual toda vez**

Quando você usa uma tupla como chave, isso funciona bem porque:

In [ ]:
(2, 3)  # sempre representa exatamente esse par

In [ ]:
(2, 3) # nunca vira outra coisa.

### O problema das listas como chave

Uma lista **pode mudar de conteúdo**, mesmo sendo “a mesma variável”:

In [ ]:
posicao = [2, 3]
# depois…
posicao[0] = 9

Agora aquela “posição” **já não representa mais o mesmo ponto**.

Para o dicionário, isso é um problema sério: ele não sabe mais onde o valor deveria estar.

Por isso, o Python **nem permite** listas como chave.

Tuplas funcionam como chaves porque **representam algo estável e confiável**, enquanto listas podem mudar e **quebrar a lógica de localização do dicionário**.

Essa regra existe para garantir que o dicionário **sempre encontre o valor certo**, rápido e sem ambiguidade.

In [ ]:
sobre_mim = {
    (0, 0): "Você está no começo, e tudo bem começar do zero",
    (1, 2): "Você já avançou, cada passo importa",
    (2, 3): "Seu progresso é real, continue acreditando em você"
}

print(sobre_mim)

### **Prática**

Crie um dicionário onde a chave é uma tupla (x, y) e o valor é uma descrição

**Dica:** Tuplas são imutáveis → podem ser chaves

In [ ]:
# Escreva seu código aqui

### **Acessando chaves e valores em um dicionário com tuplas**
### Como acessar um valor usando a chave

Para acessar um valor no dicionário, você deve informar **a chave exata**, que neste caso é uma **tupla**:

In [ ]:
sobre_mim = {
    (0, 0): "Você está no começo, e tudo bem começar do zero",
    (1, 2): "Você já avançou, cada passo importa",
    (2, 3): "Seu progresso é real, continue acreditando em você"
}

print(sobre_mim[(1, 2)])

**O que aconteceu aqui:**

* `(1, 2)` é a chave
* O Python procura essa chave no dicionário
* Retorna a frase associada a ela

### Como acessar todas as chaves

In [ ]:
sobre_mim = {
    (0, 0): "Você está no começo, e tudo bem começar do zero",
    (1, 2): "Você já avançou, cada passo importa",
    (2, 3): "Seu progresso é real, continue acreditando em você"
}

print(sobre_mim.keys())

* Retorna todas as chaves do dicionário
* Cada chave é uma tupla que representa uma posição ou estado

### Como acessar todos os valores

In [ ]:
sobre_mim = {
    (0, 0): "Você está no começo, e tudo bem começar do zero",
    (1, 2): "Você já avançou, cada passo importa",
    (2, 3): "Seu progresso é real, continue acreditando em você"
}

print(sobre_mim.values())

* Retorna todas as mensagens armazenadas no dicionário

### Como percorrer chaves e valores juntos

In [ ]:
sobre_mim = {
    (0, 0): "Você está no começo, e tudo bem começar do zero",
    (1, 2): "Você já avançou, cada passo importa",
    (2, 3): "Seu progresso é real, continue acreditando em você"
}

for posicao, mensagem in sobre_mim.items():
    print(posicao, "->", mensagem)

**O que acontece no loop:**

* `posicao` recebe a tupla (chave)
* `mensagem` recebe o texto associado a essa chave

## **Prática**

1. Acesse e imprima a mensagem associada à chave `(0, 0)`
2. Use um `for` para imprimir todas as posições e mensagens
3. Crie uma nova chave `(3, 5)` com uma frase de incentivo sua

In [ ]:
# Escreva seu código aqui

## **set**

### **O que é**

`set` é uma **coleção não ordenada**, **sem elementos duplicados**.

### **Características principais**

* Não aceita valores repetidos
* Não possui índice
* É mutável
* Muito rápido para verificações (`in`)

In [ ]:
numeros = {1, 2, 3, 3, 4}
print(numeros)

### **Principais operações**

In [ ]:
nomes_da_computacao_parte_a = {"Ada Lovelace", "Alan Turing", "Grace Hopper"}
nomes_da_computacao_parte_b = {"Grace Hopper", "Linus Torvalds", "Guido van Rossum"}

print(nomes_da_computacao_parte_a | nomes_da_computacao_parte_b)
# União: junta todos os nomes dos dois conjuntos, sem repetir

print(nomes_da_computacao_parte_a & nomes_da_computacao_parte_b)
# Interseção: mostra apenas os nomes que aparecem nos dois conjuntos

print(nomes_da_computacao_parte_a - nomes_da_computacao_parte_b)
# Diferença: mostra os nomes que estão em 'nomes_da_computacao_parte_a' mas NÃO estão em 'nomes_da_computacao_parte_b'

### **Prática**

Crie um `set` com números repetidos e observe o resultado.

**Dica:** imprima o set e compare com a lista original.

In [ ]:
# Escreva seu código aqui

### **Prática**

Receba uma lista e retorne apenas os valores únicos.

**Dica:** use `set()` e depois converta de volta para lista.

In [ ]:
# Escreva seu código aqui

### **Prática**

Compare dois conjuntos e descubra:

* valores comuns
* valores exclusivos de cada um

**Dica:** use `&`, `-` e `|`.

In [ ]:
# Escreva seu código aqui

## **enumerate**

### **O que é**

Permite percorrer uma sequência obtendo **índice + valor** ao mesmo tempo.

In [ ]:
desenhos_da_disney = ["Bela", "Tarzan", "Operação Big Hero"]

for indice, desenho in enumerate(desenhos_da_disney):
    print(indice, desenho)

O enumerate **não imprime automaticamente os elementos**

Ele cria um objeto iterável

O Python então mostra apenas onde esse objeto está na memória

Por isso aparece algo como: <enumerate at 0x7dcdc656bb50>

In [ ]:
desenhos_da_disney = ["Bela", "Tarzan", "Operação Big Hero"]
enumerate(desenhos_da_disney, start=1)

Isso é só a representação interna do objeto, não o conteúdo.

#### **Como ver os valores de verdade**

Você precisa **percorrer** ou **converter** esse objeto.

**Usando** `for`



In [ ]:
desenho_da_disney = "Operação Big Hero"

for indice, letra in enumerate(desenho_da_disney, start=1):
    print(indice, letra)

**Convertendo para lista**

In [ ]:
desenhos_da_disney = ["Bela", "Tarzan", "Operação Big Hero"]
print(list(enumerate(desenhos_da_disney, start=1)))

### **Prática**

Percorra uma lista imprimindo índice e valor.

**Dica:** use `for i, valor in enumerate()`.

In [ ]:
# Escreva seu código aqui

### **Prática**

Crie um menu numerado a partir de uma lista.

**Dica:** comece o índice em 1.

In [ ]:
# Escreva seu código aqui

### **Prática**

Encontre os índices de todos os valores pares em uma lista.

**Dica:** combine `enumerate` com `if`.

In [ ]:
# Escreva seu código aqui

## **zip**

### **O que é**

Combina múltiplas sequências elemento a elemento.

In [ ]:
filmes = ["Ação", "Romance", "Comédia", "Terror", "Animação"]
classificacao_indicativa = [16, 16, 10, 18, 0]

for filme, idade_minima in zip(filmes, classificacao_indicativa):
    print(f"Filme: {filme} | Classificação indicativa: {idade_minima}+")

### **Prática**

Combine duas listas e imprima os pares.

**Dica:** use `for a, b in zip()`.

In [ ]:
# Escreva seu código aqui

### **Prática**

Crie um dicionário a partir de duas listas.

**Dica:** use `dict(zip(chaves, valores))`.

In [ ]:
# Escreva seu código aqui

### **Prática**

Some os elementos de duas listas posição por posição.

**Dica:** use `zip` dentro de um `for`.

In [ ]:
# Escreva seu código aqui

## **any**

### **O que é**

Retorna `True` se **pelo menos um** elemento for verdadeiro.

In [ ]:
fatos_sobre_mim = [
    False,  # "Eu não consigo fazer isso"
    False,  # "Nem sempre acredito em mim"
    True    # "Eu reconheço meu próprio progresso"
]

print(any(fatos_sobre_mim))
# Retorna True porque pelo menos uma das afirmações é positiva (True)

### **Prática**

Verifique se existe algum número positivo em uma lista.

**Dica:** use compreensão de lista dentro de `any()`.

In [ ]:
# Escreva seu código aqui

### **Prática**

Verifique se alguma palavra contém mais de 5 letras.

**Dica:** `len(palavra) > 5`.

In [ ]:
# Escreva seu código aqui

### **Prática**

Verifique se algum aluno foi reprovado (nota < 6).

**Dica:** combine `any()` com listas de notas.

In [ ]:
# Escreva seu código aqui

## **all**

### **O que é**

Retorna `True` se **todos** os elementos forem verdadeiros.

In [ ]:
sobre_meu_aprendizado = [
    True,   # "Eu confio em mim"
    True,   # "Eu reconheço meu valor"
    False   # "Eu vou desistir"
]

print(all(sobre_meu_aprendizado))
# Retorna False porque nem todas as afirmações são verdadeiras (existe um False)

### **Prática**

Verifique se todos os números são positivos.

**Dica:** `all(n > 0 for n in lista)`.

In [ ]:
# Escreva seu código aqui

### **Prática**

Verifique se todas as palavras começam com letra maiúscula.

**Dica:** use `.istitle()`.

In [ ]:
# Escreva seu código aqui

### **Prática**

Valide se todas as senhas possuem no mínimo 8 caracteres.

**Dica:** use `len()` com `all()`.

In [ ]:
# Escreva seu código aqui

## **Métodos de String**

Métodos de string são funções prontas do Python que permitem manipular, analisar e transformar textos, como mudar letras, separar palavras ou verificar conteúdos.

### **Principais métodos**

```python
texto.upper()             
# Converte todos os caracteres da string para letras maiúsculas

texto.lower()             
# Converte todos os caracteres da string para letras minúsculas

texto.strip()             
# Remove espaços em branco do início e do fim do texto

texto.replace("a", "o")   
# Substitui todas as ocorrências de "a" por "o" no texto

texto.split()             
# Divide o texto em partes usando espaços como separador e retorna uma lista

texto.join(lista)         
# Junta os elementos da lista em uma única string usando 'texto' como separador

texto.startswith("A")     
# Verifica se o texto começa com a letra "A"

texto.endswith("z")       
# Verifica se o texto termina com a letra "z"

texto.isdigit()           
# Verifica se o texto contém apenas números

texto.isalpha()           
# Verifica se o texto contém apenas letras
```

### **Prática**

Transforme um texto em letras maiúsculas.

**Dica:** Lembre de criar uma variável e colocar o texto nela. Use `.upper()`.

In [ ]:
# Escreva seu código aqui

### **Prática**

Conte quantas palavras existem em uma frase.

**Dica:** Lembre de criar uma variável e colocar a frase nela. Use `.split()` e `len()`.

In [ ]:
# Escreva seu código aqui

### **Prática**

Crie um validador de senha:

* mínimo 8 caracteres
* contém número
* contém letra

**Dica:** Use `.isdigit()`, `.isalpha()`, `any()` e `all()`.

In [ ]:
# Escreva seu código aqui

##**1. List Comprehension**

### **O que é**

Forma **concisa e legível** de criar listas a partir de iteráveis.

In [ ]:
# Lista de livros
livros = ["Dom Casmurro", "O Cortiço", "Vidas Secas", "Capitães da Areia"]

# List comprehension: cria uma lista com o tamanho do nome de cada livro
tamanhos_dos_livros = [len(livro) for livro in livros]

print(tamanhos_dos_livros)
# Cada número corresponde à quantidade de caracteres do título do livro

### **Com condição**

In [ ]:
# Lista de livros
livros_carolina_maria_de_jesus = [
    "Quarto de Despejo",
    "Casa de Alvenaria",
    "Pedaços da Fome",
    "Provérbios",
    "Diário de Bitita"
]

# List comprehension com condição: seleciona apenas livros com mais de 12 caracteres no nome
livros_longos = [
    livro for livro in livros_carolina_maria_de_jesus
    if len(livro) < 12
]

print(livros_longos)
# Retorna apenas os títulos cujo nome é mais longo

['Provérbios']


### **Prática**

Crie uma lista com os seus livros favoritos

Crie uma nova lista com o **número de letras** de cada nome de livro.

**Dica:** use `len()` dentro da list comprehension.

In [ ]:
# Escreva seu código aqui

### **Prática**

Crie uma lista apenas com os livros que possuem **mais de 10 caracteres no nome**.

**Dica:** use `if` dentro da comprehension.

In [ ]:
# Escreva seu código aqui

### **Prática**

Crie uma lista com os **nomes dos livros em letras maiúsculas**, **sem espaços extras**, apenas para livros cujo nome começa com a letra **"C"**.

**Dica:** Combine `.strip()`, `.upper()` e `.startswith()` dentro da list comprehension.

In [ ]:
# Escreva seu código aqui

##**Dict Comprehension**

### **O que é**

Criação de dicionários de forma compacta.


In [ ]:
# Lista de animais de estimação
animais = ["Cachorro", "Gato", "Coelho", "Pássaro", "Hamster"]

# Dict comprehension:
# cada animal vira uma chave
# o valor é a quantidade de letras do nome do animal
animais_com_tamanho = {animal: len(animal) for animal in animais}

print(animais_com_tamanho)
# Mostra um dicionário onde cada animal está associado ao tamanho do seu nome

{'Cachorro': 8, 'Gato': 4, 'Coelho': 6, 'Pássaro': 7, 'Hamster': 7}


### **Prática**

Crie um dicionário onde:

* a **chave** seja o nome do animal selvagem
* o **valor** seja a quantidade de letras do nome

**Dica:** use `len()` dentro do dict comprehension.

In [ ]:
# Escreva seu código aqui

### **Prática**

Crie um dicionário apenas com os animais que possuem **mais de 5 letras** no nome.

**Dica:** use `if` na comprehension.

In [ ]:
# Escreva seu código aqui

### **Prática**

Inverta chaves e valores de um dicionário.

**Dica:** `{v: k for k, v in dict.items()}`

In [ ]:
# Escreva seu código aqui

##**Funções Avançadas**

### **Conceitos importantes**

* Funções como objetos
* Funções dentro de funções
* Retorno de funções

In [ ]:
def operacao(tipo):
    # função interna
    def soma(a, b):
        return a + b

    def subtracao(a, b):
        return a - b

    # escolhe qual função retornar
    if tipo == "soma":
        return soma
    elif tipo == "subtracao":
        return subtracao

# a função operacao retorna OUTRA função
somar = operacao("soma")
subtrair = operacao("subtracao")

print(somar(10, 5))
print(subtrair(10, 5))

15
5


### **Prática**

Crie uma função que retorna outra função.

**Dica:** def dentro de def.

In [ ]:
# Escreva seu código aqui

##**arg** e **kwargs**

### **O que são**

* `*args` → múltiplos argumentos posicionais
* `**kwargs` → múltiplos argumentos nomeados

In [ ]:
# *args → permite receber vários valores posicionais
def soma_pontos(*args):
    total = sum(args)
    return total


# **kwargs → permite receber vários argumentos nomeados
def info_jogador(**kwargs):
    for chave, valor in kwargs.items():
        print(f"{chave}: {valor}")


# usando *args
pontos = soma_pontos(10, 20, 15, 5)
print("Pontuação total:", pontos)

# usando **kwargs
info_jogador(
    nome="Lara",
    nivel=12,
    classe="Arqueira",
    vida=350
)

Pontuação total: 50
nome: Lara
nivel: 12
classe: Arqueira
vida: 350


### **Prática**

Crie uma função que soma qualquer quantidade de números.

**Dica:** use `*args`.

In [ ]:
# Escreva seu código aqui

### **Prática**

Crie uma função que imprime dados de um usuário.

**Dica:** use `**kwargs`.

In [ ]:
# Escreva seu código aqui

## **Funções `lambda`**

### **O que é**

`lambda` é uma forma **curta e simples** de criar funções em Python, usada quando a lógica é pequena e direta.

Pense assim:

> **`lambda` é uma função escrita em uma única linha.**

Ela funciona como uma função normal, mas:

* não usa `def`
* não usa `return`
* retorna o resultado automaticamente

* **antes dos `:`** → o que a função recebe
* **depois dos `:`** → o que a função devolve

In [ ]:
# Forma tradicional
def dobro(x):
    return x * 2

# Forma lambda
dobro = lambda x: x * 2

print(dobro(5))

10


As duas funções fazem exatamente a mesma coisa.

### **Por que usar `lambda`**

Use `lambda` quando:

* a função é **simples**
* será usada **poucas vezes**
* você quer deixar o código **mais direto**

Evite `lambda` quando:

* a lógica é grande
* há várias condições
* o código fica difícil de ler

In [ ]:
# Soma simples
soma = lambda a, b: a + b
print(soma(2, 3))  # 5

# Transformação de texto
maiusculo = lambda texto: texto.upper()
print(maiusculo("python"))  # PYTHON

# Verificação lógica
par = lambda n: n % 2 == 0
print(par(4))  # True

5
PYTHON
True


A lambda é usada apenas para definir **o que fazer com cada item**.

### **Prática**

Crie uma função lambda que:

* receba um número
* retorne o **triplo** desse número

**Dica:**
lembre-se do padrão: `entrada : saída`

In [ ]:
# Escreva seu código aqui

##**Programação Funcional**

### **Conceitos**

* Funções puras
* Imutabilidade
* Funções como argumentos

##**map**

### **O que é**

Aplica uma função a todos os elementos.

In [ ]:
numeros = [1, 2, 3, 4]

dobros = list(map(lambda x: x * 2, numeros))
print(dobros)

[2, 4, 6, 8]


### **Prática**

Use `map` para dobrar valores de uma lista.

**Dica:** map + lambda.

In [ ]:
# Escreva seu código aqui

##**filter**

### **O que é**

Filtra elementos com base em uma condição.

In [ ]:
# Lista de números positivos e negativos
numeros = [-5, -2, 0, 3, 7, -1, 10]

# filter:
# mantém apenas os números maiores que 0
positivos = list(filter(lambda x: x > 0, numeros))

print(positivos)
# O filter percorre a lista e retorna apenas os valores para os quais a função lambda devolveu True

[3, 7, 10]


### **Prática**

Filtre apenas números negativos.

**Dica:** lambda retornando True/False.

In [ ]:
# Escreva seu código aqui

## **reduce**

### **O que é**

`reduce` é uma função que **pega vários valores e combina todos eles em um único resultado final**.

Ela aplica uma função **passo a passo**, acumulando o resultado.

Pense assim:

> **reduce vai “juntando” os elementos da lista até sobrar apenas um valor.**

### **Como funciona na prática**

1. Pega os dois primeiros valores
2. Aplica a função
3. Usa o resultado junto com o próximo valor
4. Repete até acabar a sequência


In [ ]:
from functools import reduce # Veremos sobre essa linha mais para a frente

# Lista de números
numeros = [2, 3, 4, 5]

# reduce:
# multiplica todos os valores da lista
resultado = reduce(lambda acumulador, atual: acumulador * atual, numeros)

print(resultado)

120


### **O que está acontecendo**

Passo a passo:

* 2 * 3 = 6
* 6 * 4 = 24
* 24 * 5 = 120

No final, sobra **um único valor**: `120`

In [ ]:
from functools import reduce

# Pontuações de um jogador
pontuacoes = [10, 15, 20, 5]

total = reduce(lambda total, pontos: total + pontos, pontuacoes)

print(total)

O `reduce` soma todas as pontuações em um único número.

### **Prática**

Crie uma lista de números e use `reduce` para:

* calcular a soma total

**Dica:**
use `lambda acumulador, valor: acumulador + valor`

In [ ]:
# Escreva seu código aqui

##**Funções Puras**

### **O que são**

* Mesmo input → mesmo output
* Sem efeitos colaterais

In [ ]:
a = 10
b = 2

def soma(a, b):
    return a + b

soma(a, b)

12

### **Prática**

Crie uma função pura de cálculo.

**Dica:** não use print nem variáveis externas.

In [ ]:
# Escreva seu código aqui

## **Decorators**

### **O que são**

Decorators são **funções que envolvem outras funções**, adicionando comportamento **antes e/ou depois** da execução delas **sem modificar o código original**.

Pense assim:

> **Um decorator é como colocar uma “camada extra” em uma função.**

### **Por que decorators existem**

Eles servem para:

* evitar repetição de código
* adicionar funcionalidades comuns (log, validação, tempo de execução)
* deixar o código mais organizado e reutilizável



In [ ]:
# Função normal (sem decorator)
def jogar():
    print("O jogador iniciou a partida")

# Criando o decorator
def log_de_execucao(func):
    def funcao_interna():
        print("Iniciando ação...")
        func()
        print("Ação finalizada.")
    return funcao_interna

jogar = log_de_execucao(jogar)
jogar()

Iniciando ação...
O jogador iniciou a partida
Ação finalizada.


### Forma mais usada (com `@`)

In [ ]:
def log_de_execucao(func):
    def funcao_interna():
        print("Iniciando ação...")
        func()
        print("Ação finalizada.")
    return funcao_interna

@log_de_execucao #O `@log_de_execucao` é apenas uma forma mais bonita de escrever
def jogar():
    print("O jogador iniciou a partida")

jogar()

Iniciando ação...
O jogador iniciou a partida
Ação finalizada.


### O que está acontecendo por trás

1. Python lê a função `jogar`
2. Aplica o decorator `log_de_execucao`
3. Substitui `jogar` por `funcao_interna`
4. Quando você chama `jogar()`, quem roda é `funcao_interna()`

In [ ]:
def meu_decorator(func):
    def funcao_interna(nome):
        print("Olá!")
        func(nome)
        print("Até mais!")
    return funcao_interna

@meu_decorator
def saudacao(nome):
    print(f"Seja bem-vindo(a), {nome}")

saudacao("Ana")

Olá!
Seja bem-vindo(a), Ana
Até mais!


### Regra importante para não quebrar código

Use `*args` e `**kwargs` na `funcao_interna`:

* Assim, o decorator funciona com **qualquer função**.

In [ ]:
def meu_decorator(func):
    def funcao_interna(*args, **kwargs):
        print("Antes")
        resultado = func(*args, **kwargs)
        print("Depois")
        return resultado
    return funcao_interna


### **Prática**

Crie um decorator que:

* imprima `"Executando função..."`
* execute a função
* imprima `"Função finalizada"`

Aplique esse decorator a uma função que imprime uma mensagem simples.

**Dica:**
use `*args` e `**kwargs` na `funcao_interna`.

In [ ]:
# Escreva seu código aqui

##**POO – Aprofundamento**

### **Conceitos**

* Encapsulamento
* Herança
* Polimorfismo
* Métodos mágicos

##**Encapsulamento**

Encapsulamento é o conceito de **proteger os dados internos de um objeto**, controlando como eles podem ser acessados ou modificados.

Vamos pensar nos nossos jogos de vídeo game. Isso aparece para:

* impedir que a **vida do personagem seja alterada diretamente**
* garantir que regras do jogo sejam respeitadas (ex: não ter vida negativa)

Em Python, isso é feito por convenção:

* `_atributo` → uso interno
* `__atributo` → name mangling (mais protegido)

In [ ]:
class Personagem:
    def __init__(self, nome, vida):
        self.nome = nome
        self.__vida = vida  # atributo encapsulado

    def sofrer_dano(self, dano):
        if dano > 0:
            self.__vida -= dano
            if self.__vida < 0:
                self.__vida = 0

    def ver_vida(self):
        return self.__vida

O atributo `__vida` **não deve ser acessado diretamente fora da classe**, apenas por métodos controlados.

### **Prática – Encapsulamento**

Crie uma classe `Jogador` com:

* atributo privado `__energia`
* método para gastar energia
* método para recuperar energia
* método para consultar energia

**Dica:** use `__atributo` e métodos públicos para controlar o acesso.

In [ ]:
# Escreva seu código aqui

##**Herança**

Herança permite que uma classe **herde atributos e métodos de outra**, evitando repetição de código.

Em jogos, isso é muito comum:

* todos os personagens têm nome e vida
* cada tipo ataca de forma diferente

In [ ]:
class Personagem:
    def __init__(self, nome):
        self.nome = nome

    def atacar(self):
        print("O personagem ataca")

class Guerreiro(Personagem):
    def atacar(self):
        print("O guerreiro ataca com espada")

class Mago(Personagem):
    def atacar(self):
        print("O mago lança uma magia")

# Instanciando os personagens
heroi = Personagem("Genérico")
arthur = Guerreiro("Arthur")
merlin = Mago("Merlin")

# Chamando os métodos
heroi.atacar()   # Saída: O personagem ataca
arthur.atacar()  # Saída: O guerreiro ataca com espada
merlin.atacar()  # Saída: O mago lança uma magia

O personagem ataca
O guerreiro ataca com espada
O mago lança uma magia


`Guerreiro` e `Mago` **herdam de `Personagem`** e sobrescrevem o método `atacar`.

### **Prática – Herança**

Crie:

* uma classe base `Personagem`
* duas classes filhas (`Arqueiro`, `Mago` ou outras)
* cada classe deve ter um ataque diferente

**Dica:** use `class Filho(Pai)` e sobrescrita de métodos.

In [ ]:
# Escreva seu código aqui

##**Polimorfismo**

Polimorfismo significa que **objetos diferentes podem responder ao mesmo método de formas diferentes**.

Isso permite tratar todos os personagens de forma genérica.

In [ ]:
class Guerreiro:
    def atacar(self):
        return "Ataque físico"

class Mago:
    def atacar(self):
        return "Ataque mágico"

def executar_ataque(personagem):
    print(personagem.atacar())

# Criando os objetos
g1 = Guerreiro()
m1 = Mago()

# Usando a função genérica para mostrar o resultado
executar_ataque(g1)  # Saída: Ataque físico
executar_ataque(m1)  # Saída: Ataque mágico

Ataque físico
Ataque mágico


A função funciona com **qualquer personagem** que possua o método `atacar`.

### **Prática – Polimorfismo**

Crie duas classes de personagens com o mesmo método `habilidade_especial`, mas com efeitos diferentes.

**Dica:** mesmo nome de método, comportamentos distintos.

In [ ]:
# Escreva seu código aqui

##**Métodos Mágicos**

Métodos mágicos (ou *dunder methods*) começam e terminam com `__`
Eles permitem que objetos se comportem como tipos nativos do Python.

Em jogos, isso ajuda a:

* exibir personagens de forma amigável
* comparar força, nível ou itens
* usar funções como `len()` diretamente

Exemplos comuns:

* `__init__` → construtor
* `__str__` → representação em texto
* `__eq__` → comparação
* `__len__` → tamanho

In [ ]:
class Item:
    def __init__(self, nome, poder):
        self.nome = nome
        self.poder = poder

    def __str__(self):
        return f"{self.nome} (Poder: {self.poder})"

    def __eq__(self, outro):
        return self.poder == outro.poder

# Criando itens
espada = Item("Espada de Ferro", 50)
cajado = Item("Cajado Arcano", 50)
adaga = Item("Adaga Curta", 20)

# Testando o __str__ (Representação em texto)
print(f"Item 1: {espada}")
print(f"Item 2: {cajado}")

# Testando o __eq__ (Comparação de igualdade baseada no poder)
print("---")
print(f"A Espada e o Cajado têm o mesmo poder? {espada == cajado}") # True (50 == 50)
print(f"A Espada e a Adaga têm o mesmo poder? {espada == adaga}")   # False (50 == 20)

Item 1: Espada de Ferro (Poder: 50)
Item 2: Cajado Arcano (Poder: 50)
---
A Espada e o Cajado têm o mesmo poder? True
A Espada e a Adaga têm o mesmo poder? False


### **Prática – Métodos Mágicos**

Crie uma classe `Inventario` que:

* permita usar `len(inventario)` para contar itens
* permita comparar dois inventários com `==`

**Dica:** implemente `__len__` e `__eq__`.

In [ ]:
# Escreva seu código aqui

##**DESAFIO**

Crie um pequeno sistema de jogo que use:

* **encapsulamento** para proteger vida ou energia
* **herança** para diferentes tipos de personagens
* **polimorfismo** para ataques ou habilidades
* **pelo menos um método mágico**

**Dica:** pense em algo como:

* `Personagem`
* `Guerreiro`, `Mago`, `Arqueiro`
* `Item`, `Inventario`

In [ ]:
# Escreva seu código aqui

##**Tratamento de Erros**

Tratar erros é como colocar uma rede de proteção.

No mundo da programação, dividimos os erros em três categorias principais. Entender a diferença entre eles ajuda muito na hora de "debugar" (consertar) seu código.

### 1. Erros de Sintaxe (Os Erros de Gramática)

Estes acontecem antes mesmo do programa rodar. É quando você escreve algo que o Python não entende, como esquecer dois pontos (`:`) ou fechar um parêntese errado.

* **Exemplo:** Escrever `if x = 10` em vez de `if x == 10`.
* **Como tratar:** O próprio editor (como o VS Code ou o Colab) geralmente sublinha de vermelho. Basta corrigir a escrita.


### 2. Erros de Exceção (Os Erros de Execução)

O código está escrito corretamente, mas algo impossível acontece enquanto ele está rodando. Por exemplo: tentar dividir um número por zero ou abrir um arquivo que não existe.

Para esses, usamos o bloco **`try / except`**. É a forma mais comum de "tratamento de erro".

### Como funciona na prática:

In [ ]:
try:
    numero = int(input("Digite um número: "))
    resultado = 10 / numero
    print(f"Resultado: {resultado}")
except ZeroDivisionError:
    print("Erro: Você não pode dividir por zero!")
except ValueError:
    print("Erro: Por favor, digite um número válido e não uma letra.")
except Exception as e:
    print(f"Ocorreu um erro inesperado: {e}")
finally:
    print("Fim da operação.") # Isso sempre será executado

Digite um número: 5
Resultado: 2.0
Fim da operação.


### 3. Erros de Lógica (Os Erros "Silenciosos")

Esses são os mais perigosos. O programa roda perfeitamente, não exibe nenhuma mensagem de erro, mas o **resultado está errado**.

* **Exemplo:** Você quer calcular a média de duas notas, mas esquece os parênteses: `media = nota1 + nota2 / 2`. O Python vai dividir a `nota2` primeiro e depois somar, dando um valor absurdo.
* **Como tratar:**
  * **Testes Unitários:** Criar pequenos testes para verificar se as funções devolvem o que deveriam.
  * **Prints/Debug:** Colocar `print()` em partes estratégicas para ver o valor das variáveis.
  * **Assertions (`assert`):** Forçar o programa a parar se uma condição lógica for falsa.

### **Prática**

Crie tratamento para erro de divisão por zero.

**Dica:** `ZeroDivisionError`.

In [ ]:
# Escreva seu código aqui